crates/sui-core/src/transaction_manager.rs

```diff
@@ -108,7 +112,116 @@ impl Inner {
+            input_objects: HashMap::with_capacity(MIN_HASHMAP_CAPACITY),
+            pending_certificates: HashMap::with_capacity(MIN_HASHMAP_CAPACITY),
+            executing_certificates: HashMap::with_capacity(MIN_HASHMAP_CAPACITY),
+        }
+    }
+
+    // Checks if there is any transaction waiting on the lock of input_key, and try to
+    // update transactions that can acquire the lock.
+    // Must ensure input_key is available in storage before calling this function.
+    fn try_acquire_lock(&mut self, input_key: InputKey) -> Vec<PendingCertificate> {
+        let mut ready_certificates = Vec::new();
+
+        let Some(lock_queue) = self.lock_waiters.get_mut(&input_key) else {
+            // No transaction is waiting on the object yet.
+            return ready_certificates;
+        };
+
+        // Waiters can acquire lock in eitehr readonly or default mode.
+        let mut digests = BTreeSet::new();
+        if !lock_queue.readonly_waiters.is_empty() {
+            std::mem::swap(&mut digests, &mut lock_queue.readonly_waiters);
+            lock_queue.readonly_holders.extend(digests.iter().cloned());
+        } else if lock_queue.readonly_holders.is_empty() {
+            // Only try to acquire default lock if there is no readonly lock waiter / holder.
+            std::mem::swap(&mut digests, &mut lock_queue.default_waiters);
+        };
+        if lock_queue.is_empty() {
+            self.lock_waiters.remove(&input_key);
+        }
+        if digests.is_empty() {
+            return ready_certificates;
+        }
+
+        let input_count = self.input_objects.get_mut(&input_key.0).unwrap();
+        *input_count -= digests.len();
+        if *input_count == 0 {
+            self.input_objects.remove(&input_key.0);
+        }
+
+        for digest in digests {
+            // Pending certificate must exist.
+            let pending_cert = self.pending_certificates.get_mut(&digest).unwrap();
+            let lock_mode = pending_cert.acquiring_locks.remove(&input_key).unwrap();
+            assert!(pending_cert
+                .acquired_locks
+                .insert(input_key, lock_mode)
+                .is_none());
+            // When a certificate has all locks acquired, it is ready to execute.
+            if pending_cert.acquiring_locks.is_empty() {
+                let pending_cert = self.pending_certificates.remove(&digest).unwrap();
+                ready_certificates.push(pending_cert);
+            } else {
+                // TODO: we should start logging this at a higher level after some period of
+                // time has elapsed.
+                debug!(tx_digest = ?digest,acquiring = ?pending_cert.acquiring_locks, "Certificate acquiring locks");
+            }
+        }
+
+        ready_certificates
+    }
+
+    /// After reaching 3/4 load in hashmaps, increase capacity to decrease load to 1/2.
+    fn maybe_reserve_capacity(&mut self) {
```

Comment:
Do we need a MAX_HASHMAP_CAPACITY to keep this bounded? Or is it not an issue because we have MAX_PER_OBJECT_QUEUE_LENGTH & MAX_TM_QUEUE_LENGTH for protection

Reply:
I thought about this, and decided to rely on the existing limits for protection, since the actual usage at most doubles these values, which should still be tolerable in a spike.